### PR TITLE
Preserve sourcePartition to Splits mapping in task descriptor 

### PR DESCRIPTION
**Pull request by losipiuk**
## Description

We need to preserve information which splits map to which source
partition id in task descriptor to be able to split task descriptor into
multiple ones. Splitting is important in case we made a scheduling
mistake and packed to many source partitions into single task descriptor
and execution of such task is not possible due to lack of resources.

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+        public SplitsMapping build()
+        {
+            ImmutableMap.Builder<PlanNodeId, Map<Integer, List<Split>>> immutableSplitsBuilder = ImmutableMap.builder();
+            for (Map.Entry<PlanNodeId, Map<Integer, ImmutableList.Builder<Split>>> entry : splitsBuilder.entrySet()) {
+                immutableSplitsBuilder.put(entry.getKey(),
+                        entry.getValue().entrySet().stream().collect(toImmutableMap(
+                                Map.Entry::getKey,
+                                innerEntry -> innerEntry.getValue().build())));
+            }
+            return new SplitsMapping(immutableSplitsBuilder.buildOrThrow());
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java:296]: nit: this yields to streams nicely. the following woul be equally hard to follow

```
return new SplitsMapping(splitsBuilder.entrySet().stream()
                    .collect(toImmutableMap(
                            Map.Entry::getKey,
                            planNodeMapping -> planNodeMapping.getValue().entrySet().stream()
                                    .collect(toImmutableMap(
                                            Map.Entry::getKey,
                                            sourcePartitionMapping -> sourcePartitionMapping.getValue().build())))));
```

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+            implements Builder
+    {
+        private final Map<PlanNodeId, Map<Integer, ImmutableList.Builder<Split>>> splitsBuilder = new HashMap<>();
+
+        @Override
+        public Builder addSplits(PlanNodeId planNodeId, int partitionId, List<Split> splits)
+        {
+            splitsBuilder.computeIfAbsent(planNodeId, ignored -> new HashMap<>())
+                    .computeIfAbsent(partitionId, ignored -> ImmutableList.builder())
+                    .addAll(splits);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java:282]: Did you consider `Table`?

File: core/trino-main/src/test/java/io/trino/execution/scheduler/faulttolerant/SplitAssignerTester.java [hunk 3]
```diff
@@ -109,6 +119,38 @@ public void checkContainsSplits(PlanNodeId planNodeId, Collection<Split> splits,
         }
     }
 
+    public void checkContainsSplits(PlanNodeId planNodeId, ListMultimap<Integer, Split> splitsBySourcePartition, boolean replicated)
+    {
+        ListMultimap<Integer, Integer> expectedSplitIds;
+        if (replicated) {
+            expectedSplitIds = ArrayListMultimap.create();
+            expectedSplitIds.putAll(0, buildSplitIds(splitsBySourcePartition).values());
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/execution/scheduler/faulttolerant/SplitAssignerTester.java:127]: make 0 a named constant

File: core/trino-main/src/test/java/io/trino/execution/scheduler/faulttolerant/TestSplitsMapping.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalInt;
+
+import static io.trino.testing.TestingHandles.TEST_CATALOG_HANDLE;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.guava.api.Assertions.assertThat;
+
+class TestSplitsMapping
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/execution/scheduler/faulttolerant/TestSplitsMapping.java:31]: public (by convention)

**Review by findepi — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/ArbitraryDistributionSplitAssigner.java [hunk 5]
```diff
@@ -194,6 +195,13 @@ private AssignmentResult assignReplicatedSplits(PlanNodeId planNodeId, List<Spli
         return assignment.build();
     }
 
+    private ListMultimap<Integer, Split> singleSourcePartition(int sourcePartitionId, List<Split> splits)
+    {
+        ImmutableListMultimap.Builder<Integer, Split> builder = ImmutableListMultimap.builder();
+        builder.putAll(0, splits);
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/ArbitraryDistributionSplitAssigner.java:201]: USed:
```
// marker source partition id for data which is not hash distributed
    int SINGLE_SOURCE_PARTITION_ID = 0;
```

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+    private static final int INSTANCE_SIZE = instanceSize(SplitsMapping.class);
+
+    public static final SplitsMapping EMPTY = SplitsMapping.builder().build();
+
+    // not using Multimap to avoid extensive data structure copying when building updated SplitsMapping
+    private final Map<PlanNodeId, Map<Integer, List<Split>>> splits; // plan-node -> hash-partition -> Split
+
+    private SplitsMapping(ImmutableMap<PlanNodeId, Map<Integer, List<Split>>> splits)
+    {
+        this.splits = splits;
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java:52]: It is only called from builders and immutablility is enforced alrady by the type of `splits` argument. Will add a comment still

[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java:52]: Or - actually - will just do copyOf - cheap and avoids confusion.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+    }
+
+    public Set<PlanNodeId> getPlanNodeIds()
+    {
+        return splits.keySet();
+    }
+
+    public ListMultimap<PlanNodeId, Split> getSplitsFlat()
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java:60]: In the production code the iteration usecase is only for generating debug string when we run out of memory in task descriptor storage. I think it is not worth optimizing.
Let's discuss tomorrow - maybe I do not understand what benefits we can get here.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+            // TODO can we do less copying?
+            splitsFlat.putAll(entry.getKey(), entry.getValue().values().stream().flatMap(Collection::stream).collect(toImmutableList()));
+        }
+        return splitsFlat.build();
+    }
+
+    public List<Split> getSplitsFlat(PlanNodeId planNodeId)
+    {
+        Map<Integer, List<Split>> splits = this.splits.get(planNodeId);
+        if (splits == null || splits.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java:73]: Updated builders to ensure that.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+            return ImmutableList.of();
+        }
+
+        if (splits.size() == 1) {
+            return getOnlyElement(splits.values());
+        }
+
+        // TODO improve to not copy here; return view instead
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java:81]: Actually this is not important - it is test-only method. Can we enforce that method is only called from test code?

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+    {
+        return new UpdatingBuilder(mapping);
+    }
+
+    public long size()
+    {
+        return splits.values().stream().flatMap(sourcePartitionToSplits -> sourcePartitionToSplits.values().stream()).mapToLong(List::size).sum();
+    }
+
+    public interface Builder
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java:159]: nice

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+                    // reuse source immutable split list
+                    targetSplitsMapBuilder.put(sourcePartitionKey, innerEntry.getValue());
+                }
+                immutableSplitsBuilder.put(planNodeId, targetSplitsMapBuilder.build());
+            }
+
+            for (Map.Entry<PlanNodeId, Map<Integer, ImmutableList.Builder<Split>>> entry : updates.entrySet()) {
+                PlanNodeId planNodeId = entry.getKey();
+                if (originalMapping.splits.containsKey(planNodeId)) {
+                    // handled above
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java:257]: Simpler indeed.
Thanks for reminding my that I cannot code :P

[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java:257]: I think you earned the `Coauthored-By: ...`

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java (hunk 1 above)
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java:296]: indeed

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java (hunk 2 above)
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java:282]: Like
```
        Table<PlanNodeId, Integer, ImmutableList.Builder<Split>> ....;
```

It would work - but Table interface is less nice to work with. E.g you do not have `computeIfAbsent`

**Comment by losipiuk** [comment]
@findepi I added one more tiny commit - PTAL when you have a chance.

**Review by findepi — APPROVED** [review]
